### ISSUE TITLE
Wrong failing line shown on bash < 4.4

### ISSUE THREAD

**Issue by Flamefire** [bats-core/bats-core#228]
TLDR: Doing a `return 1` in a function messes up the stacktrace of bash and hence bats which will show a wrong failing line.

Example:

`assert.bash`:   
```
function fail {
  return 1
}
```
`test.bats`:
```
load assert
@test "test" {
  echo "Prev"
  # Increase linecount
  # Now fail
  fail
}
```

I traced this down to the point where bash deletes a stack entry when returning from `fail`. It (correctly) deletes the top-most entry from `$BASH_SOURCE` but the 2nd entry from `$BASH_LINENO`. This looks like this (printing the stack trace array from the debug trap):

```
[D] 18 fail assert.bash 64 test_test /tmp/bats.10668.src <- Before return: Correct
[D]18 test_test /tmp/bats.10668.src <- After return: Wrong
```
This will show a failure in the test file line 18, instead of line 64 (or line 18 in assert.bash).

I created test cases to reproduce this but was unable so far to create a fix, especially as this is fixed in bash 4.4. But this is quite surprising and very annoying to handle.
